build-go.sh

```diff
@@ -0,0 +1,237 @@
+	if [ ! $? -eq 0 ]
+	then
+		cp dist.json dist.json.temp
+		jq ".platforms[\"$goos\"] = {\"name\":\"$goos Binary\",\"archs\":{}}" dist.json.temp > dist.json
+	fi
+
+	local binname="$1"
+	if [ "$goos" = "windows" ]
+	then
+		binname="$binname.exe"
+	fi
+
+	cp dist.json dist.json.temp
+	jq ".platforms[\"$goos\"].archs[\"$goarch\"] = {\"link\":\"$goos-$goarch/$binname\"}" dist.json.temp > dist.json
+
+}
+
+function doBuild() {
+	local goos=$1
+	local goarch=$2
+	local target=$3
+	local output=$4
+
+	echo "==> building for $goos $goarch"
+
+	dir=$output/$1-$2
+	if [ -e $dir ]
+	then
+		echo "    $dir exists, skipping build"
+		return
+	fi
+	echo "    output to $dir"
+
+	mkdir -p tmp-build
+
+	(cd tmp-build && GOOS=$goos GOARCH=$goarch go build $target 2> build-log)
+	local success=$?
+	local binname=$(basename $target)
+	if [ "$success" != 0 ]
+	then
+		warn "    failed."
+		return 1
+	fi
+
+	notice "    build succeeded!"
+
+	# copy dist assets if they exist
+	if [ -e $GOPATH/src/$target/dist ]; then
+		cp -r $GOPATH/src/$target/dist/* tmp-build/
+	fi
+
+	# now zip it all up
+	mkdir -p $dir
+	if  zip -r $dir/$binname.zip tmp-build/* > /dev/null
+	then
+		printDistInfo $binname
+		rm -rf tmp-build
+	else
+		warn "    failed to zip up output"
+		success=1
+	fi
+
+
+	# output results to results table
+	echo $target, $goos, $goarch, $success >> $output/results
+}
+
+function printInitialDistfile() {
+	local distname=$1
+	local version=$2
+
+	printf "{\"id\":\"$distname\",\"version\":\"$version\",\"releaseLink\":\"releases/$distname/$version\"}" |
+	jq ".name = \"$disname\"" |
+	jq ".platforms = {}" |
+	if [  -e description ]; then
+		jq ".description = \"`cat description`\""
+	else
+		warn "no description file found"
+	fi
+}
+
+function printBuildInfo() {
+	# print out build information
+	local commit=$1
+	go version
+	echo "git sha of code: $commit" 
+	uname -a
+	echo built on `date`
+}
+
+function buildWithMatrix() {
+	local matfile=$1
+	local gobin=$2
+	local output=$3
+	local commit=$4
+
+	if [ -z "$output" ]; then
+		fail "error: output dir not specified"
+	fi
+
+	if [ ! -e $matfile ]; then
+		fail "build matrix $matfile does not exist"
+	fi
+
+	mkdir -p $output
+
+	local distname=$(basename $gobin)
+
+	printInitialDistfile $distname $version > dist.json
+	printBuildInfo $commit > $output/build-info
+
+	# build each os/arch combo
+	while read line
+	do
+		doBuild $line $gobin $output
+	done < $matfile
+
+	mv dist.json $output/dist.json
+}
+
+function cleanRepo() {
+	local repopath=$1
+
+	reporoot=$(cd $repopath && git rev-parse --show-toplevel)
+	(cd $reporoot && git clean -df)
+	(cd $reporoot && git reset --hard)
+}
+
+function checkoutVersion() {
+	local repopath=$1
+	local ref=$2
+
+	if [ -z "$repopath" ]
+	then
+		fail "checkoutVersion: no repo to check out specified"
+	fi
+
+	echo "==> checking out version $ref in $repopath"
+	cleanRepo $repopath
+	(cd $repopath && git checkout $ref > /dev/null)
```

Comment:
As paths can contain spaces especially on Windows, it's better to quote them like:

```
cleanRepo "$repopath"
(cd "$repopath" && git checkout $ref > /dev/null)
```